_posts/2021-08-09-image-resizing.md

```diff
@@ -0,0 +1,172 @@
+
+A tricky scenario that could happen, and we as the ML team experienced it, could come from the pre-processing step of a machine learning model.  
+Usually, we resize the input of a machine learning model mainly because models train faster on smaller images. An input image that is twice the size requires our network to learn from four times as many pixels, with more memory need and times that add up. 
+Moreover, many deep learning model architectures require that the input have the same size, and raw collected images might have different sizes.
+
+The workflow of the development of an ML model starts from a training phase, typically in Python. Then, if your metrics on the test set satisfy your requirements, you may want to deploy your algorithm.  
+Suppose you need to use your model in a high-performance environment such as C++, e.g., you need to integrate your model in an existing C++ application or in general. In that case, you want to use your solution in another programming language [^3], and you need a way to export "something" that could be used in the production environment.  
+A good idea to preserve the algorithm behavior is to export the whole pipeline, thus not only the forward pass of the network, given by the weights and the architecture of the layers but also the pre-and post-processing steps.
+
+Fortunately, the main deep learning frameworks, i.e., **Tensorflow** and **PyTorch**, give you the possibility to export the whole execution graph into a "program," called `SavedModel` or `TorchScript`, respectively. We used the term program because these formats include both the architecture, trained parameters, and computation.
+
+If you are developing a new model from scratch, you can design your application to export the entire pipeline, but this is not always possible if you are using a third-party library. So, for example, you can export only the inference but not the pre-processing.
+Here come the resizing problems because you probably need to use a different library to resize your input, maybe because you don't know how the rescaling is done, or there isn't the implementation of the Python library in your deploying language.  
+
+## But why the behavior of resizing is different?
+
+The definition of scaling function is mathematical and should never be a function of the library being used. Unfortunately, implementations differ across commonly-used libraries and mainly come from how it is done the **interpolation**.
+
+Image transformations are typically done in reverse order (from destination to source) to avoid sampling artifacts. 
+In practice, for each pixel $$(x,y)$$ of the destination image, you need to compute the coordinates of the corresponding pixel in the input image and copy the pixel value:
+
+$$ I_{dst}(x,y) = I_{src}\left(f_x(x,y), f_y(x,y)\right) $$
```

Comment:
If <f_x, f_y> goes from dst to src, how can this be I_dst? Shouldn't be I_src?

Reply:
<f_x, f_y> transforms output image coordinates in input image coordinates. So, pixel <x, y> in I_dst is given by the pixel <f_x(x,y), f_y(x,y)> in I_src.
I followed the same convection used [here](https://docs.opencv.org/3.4/da/d54/group__imgproc__transform.html#details)